contracts/token/ERC721/ERC721.sol

```diff
@@ -221,40 +199,73 @@ abstract contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Er
     }
 
     /**
-     * @dev Returns whether `spender` is allowed to manage `tokenId`.
-     *
-     * Requirements:
-     *
-     * - `tokenId` must exist.
+     * @dev Returns the approved address for `tokenId`. Returns 0 if `tokenId` is not minted.
      */
-    function _isApprovedOrOwner(address spender, uint256 tokenId) internal view virtual returns (bool) {
-        address owner = ownerOf(tokenId);
-        return (spender == owner || isApprovedForAll(owner, spender) || getApproved(tokenId) == spender);
+    function _getApproved(uint256 tokenId) internal view virtual returns (address) {
+        return _tokenApprovals[tokenId];
     }
 
     /**
-     * @dev Safely mints `tokenId` and transfers it to `to`.
+     * @dev Returns whether `spender` is allowed to manage `owner`'s tokens, or `tokenId` in
+     * particular (ignoring whether it is owned by `owner`).
      *
-     * Requirements:
-     *
-     * - `tokenId` must not exist.
-     * - If `to` refers to a smart contract, it must implement {IERC721Receiver-onERC721Received}, which is called upon a safe transfer.
+     * WARNING:
+     * - ownership is not checked.
+     * - spender = address(0) will lead to this function returning true in unexpected situations.
+     */
+    function _isApproved(address owner, address spender, uint256 tokenId) internal view virtual returns (bool) {
+        return owner == spender || isApprovedForAll(owner, spender) || getApproved(tokenId) == spender;
```

Comment:
Note that `getApproved` is reading the token owner because it checks existence, so we should use `_getApproved`. But then we need to protect against return value 0.

From https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4405, I believe this should be the logic for this function:

```solidity
        return (owner != address(0) &&
            (spender == owner ||
                isApprovedForAll(owner, spender) ||
                (_getApproved(tokenId) == spender && spender != address(0))));
```

Formatting is ugly, perhaps we have to refactor it into a series of if/else.

Reply:
what about 
```
function _isApproved(address owner, address spender, uint256 tokenId) internal view virtual returns (bool) {
    return
        spender != address(0) &&
        (owner == spender || isApprovedForAll(owner, spender) || _getApproved(tokenId) == spender);
}
```
?

Reply:
if owner is 0, that would return false because:
- spender must be non 0, and so `owner != spender`
- address 0 should not be able to set approval for all (that is not something we check is `_setApprovalForAll`, but we could)
- non minted token should not have an approval, so if spender is 0, `_getApproved(tokenId) != spender`

Reply:
Note: currently if you call _isApproved on with a tokenId that doesn't exist:
- the owner is fetched with `_ownerOf`, that is 0
- the owner is not the spender
- the owner has not approved the spender
- getApproved is called, and that reverts with a `ERC721NonexistentToken`

With this change the function returns false, and we get an `ERC721InsufficientApproval` returned. So to keep the errors, we'll have to add some checks higher in the call stack

Reply:
Assuming the other invariants of the contract hold (things like address(0) never approves anyone), I expect `_isAuthorized` to return false if owner is 0. So the extra `if` in `_checkAuthorized` is not a divergence from `_isAuthorized`.

Still we could rewrite it as:
```solidity
        if (!_isAuthorized(owner, spender, tokenId)) {
            if (owner == address(0)) {
                revert ERC721NonexistentToken(tokenId);
            } else {
                revert ERC721InsufficientApproval(spender, tokenId);
            }
        }
```

Reply:
I did the rewrite. For the record the previous version was
```solidity
        if (owner == address(0)) {
            revert ERC721NonexistentToken(tokenId);
        } else if (!_isAuthorized(owner, spender, tokenId)) {
            revert ERC721InsufficientApproval(spender, tokenId);
        }
```

Let me know if you prefer the previous version.